addons/mrp_subcontracting/models/stock_picking.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    display_action_record_components = fields.Boolean(compute='_compute_display_action_record_components')
+    display_view_subcontracted_move_lines = fields.Boolean(compute='_compute_display_view_subcontracted_move_lines')
+
+    @api.depends('state')
+    def _compute_display_action_record_components(self):
+        for picking in self:
+            # Hide if not encoding state
+            if picking.state in ('draft', 'cancel', 'done'):
+                continue
+            if not picking._is_subcontract():
+                continue
+            # Hide if no move is tracked
+            subcontracted_productions = picking._get_subcontracted_productions()
+            subcontracted_moves = subcontracted_productions.mapped('move_raw_ids')
+            subcontracted_moves |= subcontracted_productions.mapped('move_finished_ids')
+            if all(subcontracted_move.has_tracking == 'none' for subcontracted_move in subcontracted_moves):
+                continue
+            # Hide if the production is to close
+            if not subcontracted_productions.filtered(lambda mo: mo.state not in ('to_close', 'done')):
+                continue
+            picking.display_action_record_components = True
+
+    @api.depends('state')
+    def _compute_display_view_subcontracted_move_lines(self):
+        for picking in self:
+            # Hide if not encoding state
+            if picking.state in ('draft', 'cancel'):
+                continue
+            if not picking._is_subcontract():
+                continue
+            # Hide until state done if no move is tracked, if tracked until something was produced
+            subcontracted_productions = picking._get_subcontracted_productions()
+            subcontracted_moves = subcontracted_productions.mapped('move_raw_ids')
+            subcontracted_moves |= subcontracted_productions.mapped('move_finished_ids')
+            if all(subcontracted_move.has_tracking == 'none' for subcontracted_move in subcontracted_moves):
+                if picking.state != 'done':
+                    continue
+            # Hide if nothing was produced
+            if all(subcontracted_move.quantity_done == 0 for subcontracted_move in subcontracted_moves):
+                continue
+            picking.display_view_subcontracted_move_lines = True
+
+    # -------------------------------------------------------------------------
+    # Action methods
+    # -------------------------------------------------------------------------
+    def action_done(self):
+        for picking in self:
+            subcontracted_productions = picking.move_lines.mapped('move_orig_ids.production_id')
```

Comment:
_get_subcontracted_productions is alone and sad

Reply:
He already had two friends, but I'll give him a third one nonetheless